minimal example giving compilation error with dealii CUDA and DEAL_II_COMPILER_VECTORIZATION_LEVEL: 1

The attached minimal example (https://github.com/dealii/dealii/files/2546362/dealiiCUDAMPIProfiling.zip) gives the following compilation error when linked to dealii compiled with CUDA and DEAL_II_COMPILER_VECTORIZATION_LEVEL: 1 (SSE2). I am using the Comet machine (https://portal.xsede.org/sdsc-comet) which has Intel CPUs and P100 GPUs

```
/software/community/dftfe/dependenciesTest/installDealiiCUDAGcc6.3.0/include/deal.II/base/numbers.h(357): error: "const dealii::VectorizedArray<double> &(const dealii::VectorizedArray<double> &)" contains a vector, which is not supported in device code
          detected during:
            instantiation of class "dealii::numbers::NumberTraits<number> [with number=dealii::VectorizedArray<double>]" 
/software/community/dftfe/dependenciesTest/installDealiiCUDAGcc6.3.0/include/deal.II/base/tensor.h(125): here
            instantiation of class "dealii::Tensor<0, dim, Number> [with dim=3, Number=dealii::VectorizedArray<double>]" 
/software/community/dftfe/dependenciesTest/installDealiiCUDAGcc6.3.0/include/deal.II/base/tensor.h(427): here
            instantiation of class "dealii::Tensor<rank_, dim, Number> [with rank_=1, dim=3, Number=dealii::VectorizedArray<double>]" 
/share/apps/compute/gcc-6.3.0/include/c++/6.3.0/type_traits(671): here
            instantiation of class "std::is_trivial<_Tp> [with _Tp=dealii::Tensor<1, 3, dealii::VectorizedArray<double>>]" 
/software/community/dftfe/dependenciesTest/installDealiiCUDAGcc6.3.0/include/deal.II/base/aligned_vector.h(895): here
            instantiation of "void dealii::AlignedVector<T>::clear() [with T=dealii::Tensor<1, 3, dealii::VectorizedArray<double>>]" 
/software/community/dftfe/dependenciesTest/installDealiiCUDAGcc6.3.0/include/deal.II/base/aligned_vector.h(705): here
```

The error went away after I recompiled deaii with explicitly setting DEAL_II_COMPILER_VECTORIZATION_LEVEL to 0 in cmake/checks/check_01_cpu_features.cmake. 
cc: @masterleinad

Sorry the above close/reopen issue was by mistake.


> 
> I can't reproduce using the following CMakeLists.txt. I am using the following configuration:
> detailed.log
> In particular, I have DEAL_II_COMPILER_VECTORIZATION_LEVEL: 3.

I tried the above CMakeLists but got the following error:
_Scanning dependencies of target cudaMPIProfile
[ 33%] Building CUDA object CMakeFiles/cudaMPIProfile.dir/cudaMPIProfile.cu.o
In file included from /software/community/dftfe/dependenciesTest/dealiiContrib/dealii/include/deal.II/base/quadrature.h:20:0,
                 from /oasis/scratch/comet/dsambit/temp_project/dealiiMpiTest/cudaMPIProfile.cu:2:
/software/community/dftfe/dependenciesTest/dealiiContrib/build/include/deal.II/base/config.h:423:19: fatal error: mpi.h: No such file or directory
 #  include <mpi.h>_
Do I need to export the MPI dir path?
The CMakeLists.txt I am using is attached [here] (https://github.com/dealii/dealii/files/2704158/CMakeLists.txt).
Could you please check if the above CMakeLists.txt works for you?

> What does /software/community/dftfe/dependenciesTest/installDealiiCUDAGcc6.3.0/include/deal.II/base/numbers.h(357) contain? In a recent developer version, I see
> 
>    static DEAL_II_CUDA_HOST_DEV const number &
>                                         conjugate(const number &x);
> 
> there.

I too see the same.

`I can't find the error text "contains a vector, which is not supported in device code" anywhere in deal.II or online...`
Same here. I think the error message is coming from the nvcc compiler. I am wondering if this issue is related to the CUDA 9.2 version I am using. I have attached my 
[detailed.log](https://github.com/dealii/dealii/files/2704170/detailed.log).

--- a/include/deal.II/base/config.h.in
+++ b/include/deal.II/base/config.h.in
@@ -114,7 +114,14 @@
  */
 
 #cmakedefine DEAL_II_WORDS_BIGENDIAN
-#define DEAL_II_COMPILER_VECTORIZATION_LEVEL @DEAL_II_COMPILER_VECTORIZATION_LEVEL@
+// We need to disable SIMD vectorization for CUDA device code.
+// Otherwise, nvcc compilers from version 9 on will emit an error message like:
+// "[...] contains a vector, which is not supported in device code"
+#ifdef __CUDA_ARCH__
+#  define DEAL_II_COMPILER_VECTORIZATION_LEVEL 0
+#else
+#  define DEAL_II_COMPILER_VECTORIZATION_LEVEL @DEAL_II_COMPILER_VECTORIZATION_LEVEL@
+#endif
 #define DEAL_II_OPENMP_SIMD_PRAGMA @DEAL_II_OPENMP_SIMD_PRAGMA@
 
 

